### DISCUSSION TITLE
why upgrade to hooks?

### DISCUSSION THREAD

**Discussion by npomfret** [TanStack/table#2189] [General]
I've been trying to upgrade from react-table 6 to 7 but have found it impossible.  I'm wondering what the reasons for the internal upgrade to react hooks were.  Implementing a table feels like an order of magnitude more complex now.  Features seem to have gone missing.  The docs aren't as clear.  From a users perspective, what has improved between v6 and v7?

**Comment by efriandika** [comment]
Yeah I agree with you @npomfret ..
Until now, I am still avoiding react hooks, because of closure hell.
In my opinion, hooks is awesome and solve some cases, but I don't think hooks are readable in complex application..

lately, I can see some great react library have been migrating to hooks and end support for class component.. it makes me consider coming back to angular.. :'(

**Comment by npomfret** [comment]
Yeah, I'm sure hooks are awesome when used in the right way, but I've had to abandon list library in favour of react-bootstrap-table-next as it's so much simpler to implement.

Re-writing this lib with hooks is a fine idea if it makes the internals more elegant, but there's no reason for the user to be affected.

**Comment by Motoxpro** [comment]
@npomfret @efriandika I never used v6, but v7 is insanely powerful. From small tables to Asana tasklist style. User just handles the UI now, which makes it even _more_ so. It sounds like v6 became a burden to maintain. https://github.com/tannerlinsley/react-table/issues/1897#issuecomment-593175711

**Comment by tannerlinsley** [comment]
It's all tradeoffs. V6 was insane internally, but had a pretty great DX for drop-in solutions. However it wasn't very performant (or as good as it could have been), it was VERY difficult to style and extend, and it was a pain to maintain. V7 by contrast is headless, thus it has the ultimate flexibility in UI theming and extension and (for now) at the cost of not providing a drop-in solution like v6. V7 has become extremely easy to maintain, fix, upgrade etc, but as you said, it's all hooks now to make that possible.

Some of the "drop-in"-ness will return with time, but for now, I'm focusing on building a very strong core of performance, reliability and flexibility. Opinionated UI can either be done by you now, or by me much later after the core has been fleshed out more.
